File: backend/app/rag/retriever.py
import numpy as np
import time
from openai import OpenAI

from app.core.config import Settings
from app.db.database import SessionLocal
from app.db.repository import list_chunk_embeddings


class Retriever:
    def __init__(self, settings: Settings) -> None:
        self.settings = settings
        self.client = OpenAI(api_key=settings.active_api_key, base_url=settings.active_base_url)
        self._cache_ttl_seconds = 45.0
        self._cached_rows: list[dict] = []
        self._cache_expires_at = 0.0

    def _load_rows(self) -> list[dict]:
        now = time.time()
        if self._cached_rows and now < self._cache_expires_at:
            return self._cached_rows

        with SessionLocal() as db:
            raw_rows = list_chunk_embeddings(db)

        cached_rows: list[dict] = []
        for row in raw_rows:
            embedding = np.array(row.get("embedding", []), dtype="float32")
            if embedding.size == 0:
                continue

            vector_norm = float(row.get("vector_norm", 0.0))
            if vector_norm == 0:
                vector_norm = float(np.linalg.norm(embedding))
            if vector_norm == 0:
                continue

            cached_rows.append(
                {
                    "chunk_id": row.get("chunk_id", ""),
                    "source_url": row.get("source_url", ""),
                    "title": row.get("title", "Untitled"),
                    "text": row.get("text", ""),
                    "embedding": embedding,
                    "vector_norm": vector_norm,
                }
            )

        self._cached_rows = cached_rows
        self._cache_expires_at = now + self._cache_ttl_seconds
        return self._cached_rows

    def _embed_query(self, query: str) -> list[float]:
        request_kwargs = {
            "model": self.settings.llm_embedding_model,
            "input": query,
        }
        if self.settings.llm_embedding_query_input_type.strip():
            request_kwargs["input_type"] = self.settings.llm_embedding_query_input_type.strip()

        response = self.client.embeddings.create(**request_kwargs)
        return response.data[0].embedding



    def retrieve_with_timings(self, query: str) -> tuple[list[dict], dict[str, float]]:
        retrieval_start = time.perf_counter()
        load_start = time.perf_counter()
        rows = self._load_rows()
        row_count = len(rows)
        load_rows_ms = (time.perf_counter() - load_start) * 1000

        if not rows:
            return [], {
                "row_count": 0,
                "shape_mismatch_count": 0,
                "load_rows_ms": round(load_rows_ms, 2),
                "retrieval_ms": round((time.perf_counter() - retrieval_start) * 1000, 2),
            }

        embed_start = time.perf_counter()
        query_embedding = np.array(self._embed_query(query), dtype="float32")
        embed_query_ms = (time.perf_counter() - embed_start) * 1000
        query_norm = float(np.linalg.norm(query_embedding))
        if query_norm == 0:
            return [], {
                "load_rows_ms": round(load_rows_ms, 2),
                "embed_query_ms": round(embed_query_ms, 2),
                "retrieval_ms": round((time.perf_counter() - retrieval_start) * 1000, 2),
            }

        score_start = time.perf_counter()
        scored: list[dict] = []
        shape_mismatch_count = 0
        for row in rows:
            embedding = row.get("embedding")
            if embedding is None:
                continue
            if embedding.shape[0] != query_embedding.shape[0]:
                shape_mismatch_count += 1
                continue

            vector_norm = float(row.get("vector_norm", 0.0))
            if vector_norm == 0:
                continue

            similarity = float(np.dot(query_embedding, embedding) / (query_norm * vector_norm))
            item = {
                "chunk_id": row.get("chunk_id", ""),
                "source_url": row.get("source_url", ""),
                "title": row.get("title", "Untitled"),
                "text": row.get("text", ""),
                "score": similarity,
            }
            scored.append(item)

        score_ms = (time.perf_counter() - score_start) * 1000
        scored.sort(key=lambda item: item["score"], reverse=True)

        if not scored:
            retrieval_ms = (time.perf_counter() - retrieval_start) * 1000
            return [], {
                "row_count": row_count,
                "shape_mismatch_count": shape_mismatch_count,
                "load_rows_ms": round(load_rows_ms, 2),
                "embed_query_ms": round(embed_query_ms, 2),
                "score_ms": round(score_ms, 2),
                "retrieval_ms": round(retrieval_ms, 2),
            }

        top_hits = scored[: self.settings.top_k]
        retrieval_ms = (time.perf_counter() - retrieval_start) * 1000
        return top_hits, {
            "row_count": row_count,
            "shape_mismatch_count": shape_mismatch_count,
            "load_rows_ms": round(load_rows_ms, 2),
            "embed_query_ms": round(embed_query_ms, 2),
            "score_ms": round(score_ms, 2),
            "retrieval_ms": round(retrieval_ms, 2),
        }

    def retrieve(self, query: str) -> list[dict]:
        hits, _ = self.retrieve_with_timings(query)
        return hits

File: backend/src/rag/retriever.ts
import OpenAI from "openai";
import { performance } from "perf_hooks";
import { getSettings } from "../core/config";
import { listChunkEmbeddings } from "../db/repository";
import logger from "../core/logger";

interface CachedRow {
  chunk_id: string;
  source_url: string;
  title: string;
  text: string;
  embedding: Float32Array;
  vector_norm: number;
}

interface RetrievalResult {
  chunk_id: string;
  source_url: string;
  title: string;
  text: string;
  score: number;
}

export class Retriever {
  private client: OpenAI;
  private settings: any;
  private cachedRows: CachedRow[] = [];
  private cacheExpiresAt = 0;
  private cacheTtlSeconds = 45.0;

  constructor() {
    this.settings = getSettings();
    this.client = new OpenAI({
      apiKey: this.settings.getActiveApiKey(),
      baseURL: this.settings.getActiveBaseUrl(),
    });
  }

  private async loadRows(): Promise<CachedRow[]> {
    const now = Date.now() / 1000;
    if (this.cachedRows.length > 0 && now < this.cacheExpiresAt) {
      return this.cachedRows;
    }

    const rawRows = await listChunkEmbeddings();

    const cachedRows: CachedRow[] = [];
    for (const row of rawRows) {
      const embedding = new Float32Array(row.embedding || []);

      if (embedding.length === 0) {
        continue;
      }

      let vectorNorm = row.vectorNorm || 0;
      if (vectorNorm === 0) {
        vectorNorm = this.norm(embedding);
      }
      if (vectorNorm === 0) {
        continue;
      }

      cachedRows.push({
        chunk_id: row.chunkId,
        source_url: row.sourceUrl,
        title: row.title,
        text: row.text,
        embedding,
        vector_norm: vectorNorm,
      });
    }

    this.cachedRows = cachedRows;
    this.cacheExpiresAt = now + this.cacheTtlSeconds;
    return this.cachedRows;
  }

  private norm(vec: Float32Array): number {
    let sum = 0;
    for (let i = 0; i < vec.length; i++) {
      sum += vec[i] * vec[i];
    }
    return Math.sqrt(sum);
  }

  private dotProduct(
    a: Float32Array,
    b: Float32Array
  ): number {
    let sum = 0;
    for (let i = 0; i < a.length; i++) {
      sum += a[i] * b[i];
    }
    return sum;
  }

  private async embedQuery(query: string): Promise<number[]> {
    const requestParams: any = {
      model: this.settings.llmEmbeddingModel,
      input: query,
    };

    if (this.settings.llmEmbeddingQueryInputType) {
      requestParams.input_type =
        this.settings.llmEmbeddingQueryInputType;
    }

    const response = await this.client.embeddings.create(requestParams);
    return response.data[0].embedding;
  }

  async retrieveWithTimings(
    query: string
  ): Promise<[RetrievalResult[], Record<string, any>]> {
    const retrievalStart = performance.now();

    const loadStart = performance.now();
    const rows = await this.loadRows();
    const rowCount = rows.length;
    const loadRowsMs = performance.now() - loadStart;

    if (rows.length === 0) {
      return [
        [],
        {
          row_count: 0,
          shape_mismatch_count: 0,
          load_rows_ms: Math.round(loadRowsMs * 100) / 100,
          retrieval_ms: Math.round((performance.now() - retrievalStart) * 100) / 100,
        },
      ];
    }

    const embedStart = performance.now();
    const queryEmbedding = new Float32Array(await this.embedQuery(query));
    const embedQueryMs = performance.now() - embedStart;

    const queryNorm = this.norm(queryEmbedding);
    if (queryNorm === 0) {
      return [
        [],
        {
          load_rows_ms: Math.round(loadRowsMs * 100) / 100,
          embed_query_ms: Math.round(embedQueryMs * 100) / 100,
          retrieval_ms: Math.round((performance.now() - retrievalStart) * 100) / 100,
        },
      ];
    }

    const scoreStart = performance.now();
    const scored: Array<RetrievalResult & { raw_score: number }> = [];
    let shapeMismatchCount = 0;

    for (const row of rows) {
      const embedding = row.embedding;
      if (!embedding) {
        continue;
      }
      if (embedding.length !== queryEmbedding.length) {
        shapeMismatchCount++;
        continue;
      }

      const dotProd = this.dotProduct(embedding, queryEmbedding);
      const score = dotProd / (row.vector_norm * queryNorm);

      scored.push({
        chunk_id: row.chunk_id,
        source_url: row.source_url,
        title: row.title,
        text: row.text,
        score,
        raw_score: dotProd,
      });
    }

    scored.sort((a, b) => b.raw_score - a.raw_score);

    const topK = Math.min(this.settings.topK, scored.length);
    const hits = scored.slice(0, topK).map(({ raw_score, ...rest }) => rest);

    const scoreMs = performance.now() - scoreStart;

    return [
      hits,
      {
        row_count: rowCount,
        shape_mismatch_count: shapeMismatchCount,
        load_rows_ms: Math.round(loadRowsMs * 100) / 100,
        embed_query_ms: Math.round(embedQueryMs * 100) / 100,
        score_ms: Math.round(scoreMs * 100) / 100,
        retrieval_ms: Math.round((performance.now() - retrievalStart) * 100) / 100,
      },
    ];
  }
}
